// appium-ocr-plugin/build/test/unit/plugin-specs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("../..");
const earl_1 = require("earl");
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const SETTINGS = {};
const SCREEN_FIXTURE = path_1.default.resolve(__dirname, '..', '..', '..', 'fixtures', 'screen.png');
function getFakeDriver({ settings = SETTINGS } = {}) {
    return {
        settings: {
            getSettings() {
                return settings;
            },
        },
        opts: {
            platformName: 'iOS',
        },
        getScreenshot() {
            return __awaiter(this, void 0, void 0, function* () {
                const screen = yield promises_1.default.readFile(SCREEN_FIXTURE);
                return screen.toString('base64');
            });
        }
    };
}
describe('AppiumOcrPlugin', function () {
    it('should ready a worker', function () {
        return __awaiter(this, void 0, void 0, function* () {
            (0, earl_1.expect)(__1.AppiumOcrPlugin).toBeTruthy();
            const p = new __1.AppiumOcrPlugin('foo');
            (0, earl_1.expect)(p.isWorkerReady).toEqual(false);
            const d = getFakeDriver();
            yield p.readyWorker(d);
            (0, earl_1.expect)(p.isWorkerReady).toEqual(true);
            if (p.worker) {
                yield p.worker.terminate();
            }
        });
    });
    it('should process a screenshot', function () {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const p = new __1.AppiumOcrPlugin('test');
            const d = getFakeDriver();
            const { words } = yield p.getOcrText(() => __awaiter(this, void 0, void 0, function* () { }), d);
            const echoBoxWords = words.filter((w) => w.text === 'EchoBox');
            const echoBoxWord = echoBoxWords[0];
            (0, earl_1.expect)(echoBoxWord).toBeTruthy();
            (0, earl_1.expect)(echoBoxWord.confidence).toBeGreaterThan(70);
            yield ((_a = p.worker) === null || _a === void 0 ? void 0 : _a.terminate());
        });
    });
});
