// appium-ocr-plugin/build/commands/element.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAttribute = exports.getText = exports.getElementRect = exports.getLocation = exports.getSize = exports.elementDisplayed = exports.click = exports.ocrElementGuard = exports.OCRElement = exports.OCR_ELEMENT_PREFIX = void 0;
const support_1 = require("appium/support");
const driver_1 = require("appium/driver");
const TAP_DURATION_MS = 250;
exports.OCR_ELEMENT_PREFIX = 'ocr-element-';
class OCRElement {
    constructor(text, confidence, bbox) {
        this.text = text;
        this.confidence = confidence;
        this.rect = {
            x: bbox.x0,
            y: bbox.y0,
            width: bbox.x1 - bbox.x0,
            height: bbox.y1 - bbox.y0,
        };
        this.id = `${exports.OCR_ELEMENT_PREFIX}${support_1.util.uuidV4()}`;
    }
    get size() {
        return { width: this.rect.width, height: this.rect.height };
    }
    get location() {
        return { x: this.rect.x, y: this.rect.y };
    }
    get center() {
        return {
            x: this.rect.x + this.rect.width / 2,
            y: this.rect.y + this.rect.height / 2,
        };
    }
    get asW3CElementObject() {
        return { [support_1.util.W3C_WEB_ELEMENT_IDENTIFIER]: this.id };
    }
}
exports.OCRElement = OCRElement;
function ocrElementGuard(next, elId, fn) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrContextGuard(next, () => __awaiter(this, void 0, void 0, function* () {
            const el = this.ocrElements[elId];
            if (!el) {
                throw new driver_1.errors.NoSuchElementError();
            }
            return yield fn(el);
        }));
    });
}
exports.ocrElementGuard = ocrElementGuard;
function click(next, driver, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () {
            const { x, y } = el.center;
            this.logger.info(`Will tap on image element at coordinate [${x}, ${y}]`);
            const action = {
                type: 'pointer',
                id: 'mouse',
                parameters: { pointerType: 'touch' },
                actions: [
                    { type: 'pointerMove', x, y, duration: 0 },
                    { type: 'pointerDown', button: 0 },
                    { type: 'pause', duration: TAP_DURATION_MS },
                    { type: 'pointerUp', button: 0 },
                ]
            };
            // check if the driver has the appropriate performActions method
            if (driver.performActions) {
                return yield driver.performActions([action]);
            }
            throw new Error("Driver did not implement the 'performActions' command");
        }));
    });
}
exports.click = click;
function elementDisplayed(next, _, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, () => __awaiter(this, void 0, void 0, function* () { return true; }));
    });
}
exports.elementDisplayed = elementDisplayed;
function getSize(next, _, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () { return el.size; }));
    });
}
exports.getSize = getSize;
function getLocation(next, _, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () { return el.location; }));
    });
}
exports.getLocation = getLocation;
function getElementRect(next, _, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () { return el.rect; }));
    });
}
exports.getElementRect = getElementRect;
function getText(next, _, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () { return el.text; }));
    });
}
exports.getText = getText;
function getAttribute(next, _, attr, elId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrElementGuard(next, elId, (el) => __awaiter(this, void 0, void 0, function* () {
            if (attr !== 'confidence') {
                throw new Error(`Unsupported OCR element attribute '${attr}'`);
            }
            return el.confidence;
        }));
    });
}
exports.getAttribute = getAttribute;
