// appium-ocr-plugin/build/commands/find.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._find = exports.findElements = exports.findElement = void 0;
const xpath_1 = require("xpath");
const xmldom_1 = require("xmldom");
const element_1 = require("./element");
// for some reason tsc can't find the errors export from basedriver, but it's there
const bdStar = __importStar(require("appium/driver"));
const { errors } = bdStar;
function findElement(next, driver, strategy, selector) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this._find(next, driver, strategy, selector, false);
    });
}
exports.findElement = findElement;
function findElements(next, driver, strategy, selector) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this._find(next, driver, strategy, selector, true);
    });
}
exports.findElements = findElements;
function _find(next, driver, strategy, selector, multiple) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield this.ocrContextGuard(next, () => __awaiter(this, void 0, void 0, function* () {
            if (strategy !== 'xpath') {
                throw new errors.InvalidArgumentError(`The OCR context only supports the 'xpath' locator strategy`);
            }
            const xmlStr = yield this.getPageSource(next, driver);
            const dom = new xmldom_1.DOMParser().parseFromString(xmlStr);
            const nodes = (0, xpath_1.select)(selector, dom);
            const transformNodeToOcrElement = (n) => {
                const bbox = {
                    x0: parseFloat(getNodeAttrVal(n, 'x0')),
                    y0: parseFloat(getNodeAttrVal(n, 'y0')),
                    x1: parseFloat(getNodeAttrVal(n, 'x1')),
                    y1: parseFloat(getNodeAttrVal(n, 'y1')),
                };
                const confidence = parseFloat(getNodeAttrVal(n, 'confidence'));
                const text = n.firstChild.data;
                const ocrEl = new element_1.OCRElement(text, confidence, bbox);
                this.ocrElements[ocrEl.id] = ocrEl;
                this.logger.info(`Found new element '${ocrEl.id}': [${text}][${confidence}[${JSON.stringify(ocrEl.rect)}]`);
                return ocrEl.asW3CElementObject;
            };
            const els = nodes.filter(isOcrElementNodeValid).map(transformNodeToOcrElement);
            if (!multiple) {
                if (els.length < 1) {
                    throw new errors.NoSuchElementError();
                }
                return els[0];
            }
            return els;
        }));
    });
}
exports._find = _find;
function isOcrElementNodeValid(n) {
    if (n.nodeName !== 'item') {
        return false;
    }
    if (!n.firstChild) {
        return false;
    }
    if (!n.firstChild.data) {
        return false;
    }
    return true;
}
function getNodeAttrVal(node, attr) {
    const attrObjs = Object.values(node.attributes).filter((obj) => obj.name === attr);
    if (!attrObjs.length) {
        throw new Error(`Tried to retrieve a node attribute '${attr}' but the node didn't have it`);
    }
    return attrObjs[0].value;
}
