// appium-altunity-plugin/build/src/client/commands/actions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clickElement = exports.tapElement = exports.tapCoordinates = exports.pressKey = exports.waitTillActionFinished = exports.isActionFinished = exports.keyUp = exports.keyDown = void 0;
const bluebird_1 = __importDefault(require("bluebird"));
const __1 = require("..");
const DEFAULT_TAP_COUNT = 1;
const DEFAULT_TAP_INTERVAL_MS = 250;
const WAIT_VALIDATIONS = [__1.DEFAULT_VALIDATION_STR, 'Finished'];
async function keyDown(keyCode, power = 1) {
    return await this.sendSimpleCommand('keyDown', { keyCode, power });
}
exports.keyDown = keyDown;
async function keyUp(keyCode) {
    return await this.sendSimpleCommand('keyUp', { keyCode });
}
exports.keyUp = keyUp;
async function isActionFinished() {
    const res = await this.sendSimpleCommand('actionFinished');
    return res === 'Yes';
}
exports.isActionFinished = isActionFinished;
async function waitTillActionFinished(durationMsHint) {
    await bluebird_1.default.delay(durationMsHint);
    const interval = 300;
    while (!await this.isActionFinished()) {
        await bluebird_1.default.delay(interval);
    }
}
exports.waitTillActionFinished = waitTillActionFinished;
async function pressKey(keyCode, durationMs, power = 1) {
    const durationSecs = durationMs / 1000;
    await this.sendTwoPartCommand('pressKeyboardKey', {
        keyCode,
        power,
        duration: durationSecs,
        wait: true,
    }, WAIT_VALIDATIONS);
}
exports.pressKey = pressKey;
function getTapData(args, wait, element) {
    args.count ?? (args.count = DEFAULT_TAP_COUNT);
    args.intervalMs ?? (args.intervalMs = DEFAULT_TAP_INTERVAL_MS);
    const intervalSecs = args.intervalMs / 1000;
    const data = {
        coordinates: args.pos,
        count: args.count,
        interval: intervalSecs,
        wait,
    };
    if (element) {
        data.altUnityObject = element.asUnityObject();
    }
    return data;
}
async function tapCoordinates(args, wait = true) {
    if (!args.pos) {
        throw new Error('Position is required for tap');
    }
    const data = getTapData(args, wait);
    if (wait) {
        return await this.sendTwoPartCommand('tapCoordinates', data, WAIT_VALIDATIONS);
    }
    return await this.sendSimpleCommand('tapCoordinates', data);
}
exports.tapCoordinates = tapCoordinates;
async function tapElement(element, args, wait = true, cmdName = 'tapElement') {
    const data = getTapData(args, wait, element);
    if (wait) {
        return await this.sendTwoPartCommand(cmdName, data, WAIT_VALIDATIONS);
    }
    return await this.sendSimpleCommand(cmdName, data);
}
exports.tapElement = tapElement;
async function clickElement(element, args, wait = true) {
    return await this.tapElement(element, args, wait, 'clickElement');
}
exports.clickElement = clickElement;
