// appium-altunity-plugin/build/src/client/commands/general.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTimeScale = exports.getTimeScale = exports.loadScene = exports.getCurrentScene = exports.getScreenshotAsPNG = exports.getScreenshotAsB64 = exports.getServerVersion = void 0;
const alt_element_1 = __importDefault(require("../alt-element"));
async function getServerVersion() {
    return await this.sendSimpleCommand('getServerVersion');
}
exports.getServerVersion = getServerVersion;
async function getScreenshotAsB64() {
    return await this.sendTwoPartCommand('getPNGScreenshot');
}
exports.getScreenshotAsB64 = getScreenshotAsB64;
async function getScreenshotAsPNG() {
    const b64Str = await this.getScreenshotAsB64();
    return Buffer.from(b64Str, 'base64');
}
exports.getScreenshotAsPNG = getScreenshotAsPNG;
async function getCurrentScene() {
    const res = await this.sendSimpleCommand('getCurrentScene');
    return (new alt_element_1.default(this, res)).name;
}
exports.getCurrentScene = getCurrentScene;
async function loadScene(sceneName, loadSingle = true) {
    await this.sendTwoPartCommand('loadScene', { sceneName, loadSingle }, ['Ok', 'Scene Loaded']);
}
exports.loadScene = loadScene;
async function getTimeScale() {
    const timeScale = await this.sendSimpleCommand('getTimeScale');
    return parseFloat(timeScale);
}
exports.getTimeScale = getTimeScale;
async function setTimeScale(timeScale) {
    await this.sendSimpleCommand('setTimeScale', { timeScale: timeScale.toString() });
}
exports.setTimeScale = setTimeScale;
