// appium-altunity-plugin/build/client/commands/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setComponentProperty = exports.getComponentProperty = void 0;
async function getComponentProperty(element, component, property, assemblyName, maxDepth = 2) {
    const componentName = typeof component === 'string' ? component : component.name;
    return await this.sendSimpleCommand('getObjectComponentProperty', {
        altUnityObject: element.asUnityObject(),
        component: componentName,
        property: property,
        assembly: assemblyName,
        maxDepth,
    });
}
exports.getComponentProperty = getComponentProperty;
async function setComponentProperty(element, component, property, assemblyName, value) {
    const componentName = typeof component === 'string' ? component : component.name;
    return await this.sendSimpleCommand('setObjectComponentProperty', {
        altUnityObject: element.asUnityObject(),
        component: componentName,
        property: property,
        assembly: assemblyName,
        value: JSON.stringify(value),
    });
}
exports.setComponentProperty = setComponentProperty;
