// appium-altunity-plugin/build/client/commands/find.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHierarchy = exports.getCurrentHierarchy = exports.findObject = exports.findAllObjects = exports.findObjects = void 0;
const by_1 = require("../by");
const command_helpers_1 = require("./command-helpers");
const alt_element_1 = __importDefault(require("../alt-element"));
async function findObjects(params) {
    const path = (0, command_helpers_1.getPath)(params.by, params.selector, params.contains);
    const cameraPath = (0, command_helpers_1.getPath)(this.cameraBy, this.cameraPath);
    const res = await this.sendSimpleCommand('findObjects', {
        by: params.by,
        path,
        cameraBy: this.cameraBy,
        cameraPath,
        enabled: params.enabled,
    });
    return res.map((data) => new alt_element_1.default(this, data));
}
exports.findObjects = findObjects;
async function findAllObjects(enabled = true) {
    return await this.findObjects({ by: by_1.AltBy.PATH, selector: '//*', enabled });
}
exports.findAllObjects = findAllObjects;
async function findObject(params) {
    const enabled = params.enabled ?? true;
    const path = (0, command_helpers_1.getPath)(params.by, params.selector, params.contains);
    const res = await this.sendSimpleCommand('findObject', {
        by: params.by,
        path,
        cameraBy: this.cameraBy,
        cameraPath: (0, command_helpers_1.getPath)(this.cameraBy, this.cameraPath),
        enabled,
    });
    return new alt_element_1.default(this, res);
}
exports.findObject = findObject;
function buildHierarchyFromElement(element, allEls) {
    // find all the children of this element
    const children = allEls
        .filter((child) => child.transformParentId === element.transformId)
        .map((child) => buildHierarchyFromElement(child, allEls));
    return {
        element,
        children
    };
}
async function getCurrentHierarchy() {
    const els = await this.findAllObjects();
    return await this.getHierarchy(els);
}
exports.getCurrentHierarchy = getCurrentHierarchy;
async function getHierarchy(els) {
    const rootEls = els.filter((e) => e.transformParentId === 0);
    return rootEls.map((e) => buildHierarchyFromElement(e, els));
}
exports.getHierarchy = getHierarchy;
