// appium-altunity-plugin/build/src/client/alt-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AltElement = exports.ALT_ELEMENT_KEYS = void 0;
// XXX sad that we have to duplicate these because we can't generate the type from this or this
// from the type :-(
exports.ALT_ELEMENT_KEYS = [
    'name',
    'id',
    'x',
    'y',
    'z',
    'mobileY',
    'type',
    'enabled',
    'worldX',
    'worldY',
    'worldZ',
    'idCamera',
    'transformParentId',
    'transformId',
];
class AltElement {
    constructor(client, data) {
        this.client = client;
        this.data = data;
    }
    get name() { return this.data.name; }
    get id() { return this.data.id; }
    get x() { return this.data.x; }
    get y() { return this.data.y; }
    get z() { return this.data.z; }
    get mobileY() { return this.data.mobileY; }
    get type() { return this.data.type; }
    get enabled() { return this.data.enabled; }
    get worldX() { return this.data.worldX; }
    get worldY() { return this.data.worldY; }
    get worldZ() { return this.data.worldZ; }
    get idCamera() { return this.data.idCamera; }
    get transformParentId() { return this.data.transformParentId; }
    get transformId() { return this.data.transformId; }
    asUnityObject() {
        return this.data;
    }
    toJSON() {
        return JSON.stringify(this.data);
    }
    toXMLNode() {
        return `<${this.name} id="${this.id}" x="${this.x}" y="${this.y}" type="${this.type}" parentId="${this.transformParentId}" transformId="${this.transformId}" />`;
    }
    async getText() {
        return await this.client.getElementText(this);
    }
    async getComponents() {
        return await this.client.getElementComponents(this);
    }
    async tap(args = {}) {
        return await this.client.tapElement(this, args);
    }
    async tapViaCoords() {
        return await this.client.tapCoordinates({ pos: { x: this.x, y: this.y } });
    }
    async click(args = {}) {
        return await this.client.clickElement(this, args);
    }
    async getComponentProperty(component, property, assemblyName, maxDepth) {
        return await this.client.getComponentProperty(this, component, property, assemblyName, maxDepth);
    }
    async setComponentProperty(component, property, assemblyName, value) {
        return await this.client.setComponentProperty(this, component, property, assemblyName, value);
    }
}
exports.default = AltElement;
exports.AltElement = AltElement;
