// appium-gestures-plugin/lib/plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_plugin_1 = __importDefault(require("@appium/base-plugin"));
const dragAndDrop_1 = __importDefault(require("./gestures/dragAndDrop"));
const swipe_1 = require("./gestures/swipe");
const doubleTap_1 = __importDefault(require("./gestures/doubleTap"));
const longPress_1 = __importDefault(require("./gestures/longPress"));
class GesturesPlugin extends base_plugin_1.default {
    static executeMethodMap = {
        'gesture: dragAndDrop': {
            command: 'dragAndDrop',
            params: { required: ['sourceId', 'destinationId'] },
        },
        'gesture: swipe': {
            command: 'swipe',
            params: {
                required: ['elementId', 'percentage', 'direction'],
            },
        },
        'gesture: scrollElementIntoView': {
            command: 'scrollElementIntoView',
            params: {
                required: ['scrollableView', 'strategy', 'selector', 'percentage', 'direction', 'maxCount'],
            },
        },
        'gesture: doubleTap': {
            command: 'doubleTap',
            params: { required: ['elementId'] },
        },
        'gesture: longPress': {
            command: 'longPress',
            params: { required: ['elementId', 'pressure', 'duration'] },
        },
    };
    constructor(pluginName) {
        super(pluginName);
    }
    async execute(next, driver, script, args) {
        return await this.executeMethod(next, driver, script, args);
    }
    async swipe(next, driver, elementId, percentage, direction) {
        await (0, swipe_1.swipe)(elementId, percentage, direction, driver);
    }
    async scrollElementIntoView(next, driver, scrollableView, strategy, selector, percentage, direction, maxCount) {
        await (0, swipe_1.scrollElementIntoView)({
            scrollableView,
            strategy,
            selector,
            percentage,
            direction,
            maxCount,
            driver,
        });
    }
    async dragAndDrop(next, driver, sourceId, destinationId) {
        await (0, dragAndDrop_1.default)(sourceId, destinationId, driver);
    }
    async doubleTap(next, driver, elementId) {
        await (0, doubleTap_1.default)(elementId, driver);
    }
    async longPress(next, driver, elementId, pressure, duration) {
        await (0, longPress_1.default)(elementId, pressure, duration, driver);
    }
}
exports.default = GesturesPlugin;
