// appium-altunity-plugin/build/client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_ALTUNITY_PORT = exports.DEFAULT_ALTUNITY_HOST = exports.AltKeyCode = exports.AltElement = exports.AltBy = exports.AltUnityClient = exports.AltUnityError = exports.DEFAULT_VALIDATION_STR = void 0;
const commands_1 = require("./commands");
const by_1 = require("./by");
Object.defineProperty(exports, "AltBy", { enumerable: true, get: function () { return by_1.AltBy; } });
const key_code_1 = require("./key-code");
Object.defineProperty(exports, "AltKeyCode", { enumerable: true, get: function () { return key_code_1.AltKeyCode; } });
const alt_element_1 = require("./alt-element");
Object.defineProperty(exports, "AltElement", { enumerable: true, get: function () { return alt_element_1.AltElement; } });
const connection_1 = require("./connection");
const asyncbox_1 = require("asyncbox");
const DEFAULT_ALTUNITY_HOST = '127.0.0.1';
exports.DEFAULT_ALTUNITY_HOST = DEFAULT_ALTUNITY_HOST;
const DEFAULT_ALTUNITY_PORT = 13000;
exports.DEFAULT_ALTUNITY_PORT = DEFAULT_ALTUNITY_PORT;
exports.DEFAULT_VALIDATION_STR = 'Ok';
var ResponseEncoding;
(function (ResponseEncoding) {
    ResponseEncoding["START"] = "altstart::";
    ResponseEncoding["RESPONSE"] = "::response::";
    ResponseEncoding["LOG"] = "::altLog::";
    ResponseEncoding["END"] = "::altend";
})(ResponseEncoding || (ResponseEncoding = {}));
var RequestEncoding;
(function (RequestEncoding) {
    RequestEncoding["SEPARATOR"] = ";";
    RequestEncoding["END"] = "&";
})(RequestEncoding || (RequestEncoding = {}));
class AltUnityError extends Error {
    constructor(type, message, trace) {
        super(message);
        this.type = type;
        this.trace = trace;
    }
}
exports.AltUnityError = AltUnityError;
class AltUnityClient {
    constructor(opts) {
        // general commands
        this.getServerVersion = commands_1.getServerVersion;
        this.getScreenshotAsPNG = commands_1.getScreenshotAsPNG;
        this.getScreenshotAsB64 = commands_1.getScreenshotAsB64;
        this.getCurrentScene = commands_1.getCurrentScene;
        this.loadScene = commands_1.loadScene;
        this.getTimeScale = commands_1.getTimeScale;
        this.setTimeScale = commands_1.setTimeScale;
        // find commands
        this.findObject = commands_1.findObject;
        this.findObjects = commands_1.findObjects;
        this.findAllObjects = commands_1.findAllObjects;
        this.getCurrentHierarchy = commands_1.getCurrentHierarchy;
        this.getHierarchy = commands_1.getHierarchy;
        // action commands
        this.keyDown = commands_1.keyDown;
        this.keyUp = commands_1.keyUp;
        this.pressKey = commands_1.pressKey;
        this.isActionFinished = commands_1.isActionFinished;
        this.waitTillActionFinished = commands_1.waitTillActionFinished;
        this.tapCoordinates = commands_1.tapCoordinates;
        this.tapElement = commands_1.tapElement;
        this.clickElement = commands_1.clickElement;
        // element commands
        this.getElementText = commands_1.getElementText;
        this.getElementComponents = commands_1.getElementComponents;
        // component commands
        this.getComponentProperty = commands_1.getComponentProperty;
        this.setComponentProperty = commands_1.setComponentProperty;
        if (!opts.host) {
            opts.host = DEFAULT_ALTUNITY_HOST;
        }
        if (!opts.port) {
            opts.port = DEFAULT_ALTUNITY_PORT;
        }
        this.host = opts.host;
        this.port = opts.port;
        this.conn = new connection_1.Connection({
            log: opts.log,
            port: opts.port,
            host: opts.host
        });
        this.log = opts.log;
        this.cameraBy = by_1.AltBy.NAME;
        this.cameraPath = ''; // default camera path is empty
    }
    async connect() {
        let numTries = 0;
        const maxTries = 15;
        await (0, asyncbox_1.retryInterval)(maxTries, 1000, async () => {
            numTries++;
            this.log.info(`Attempting to connect to AltUnity server. This is attempt ${numTries}/${maxTries}`);
            try {
                await this.conn.connect();
            }
            catch (err) {
                throw new Error(`Could not connect to the AltUnity server. Original error was: ${err.message}`);
            }
        });
        this.log.info(`Connection to AltUnity server established`);
    }
    async disconnect() {
        await this.conn.close();
    }
    async _sendCommand(commandName, data, responseCount = 1, validations = []) {
        const curMsgId = Date.now();
        const message = { ...data, commandName, messageId: curMsgId.toString() };
        const responses = await this.conn.sendMessage(message, responseCount, validations);
        for (const res of responses) {
            if (res.error) {
                throw new AltUnityError(res.error.type, res.error.message, res.error.trace);
            }
        }
        return responses.map((r) => r.data);
    }
    async sendSimpleCommand(commandName, data) {
        const datas = await this._sendCommand(commandName, data, 1);
        return datas[0];
    }
    async sendTwoPartCommand(commandName, data, validations = [exports.DEFAULT_VALIDATION_STR]) {
        const datas = await this._sendCommand(commandName, data, 2, validations);
        return datas[1];
    }
}
exports.default = AltUnityClient;
exports.AltUnityClient = AltUnityClient;
