// appium-altunity-plugin/build/client/key-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AltKeyCode = void 0;
var AltKeyCode;
(function (AltKeyCode) {
    AltKeyCode[AltKeyCode["NoKey"] = 0] = "NoKey";
    AltKeyCode[AltKeyCode["Backspace"] = 8] = "Backspace";
    AltKeyCode[AltKeyCode["Tab"] = 9] = "Tab";
    AltKeyCode[AltKeyCode["Clear"] = 12] = "Clear";
    AltKeyCode[AltKeyCode["Return"] = 13] = "Return";
    AltKeyCode[AltKeyCode["Pause"] = 19] = "Pause";
    AltKeyCode[AltKeyCode["Escape"] = 27] = "Escape";
    AltKeyCode[AltKeyCode["Space"] = 32] = "Space";
    AltKeyCode[AltKeyCode["Exclaim"] = 33] = "Exclaim";
    AltKeyCode[AltKeyCode["DoubleQuote"] = 34] = "DoubleQuote";
    AltKeyCode[AltKeyCode["Hash"] = 35] = "Hash";
    AltKeyCode[AltKeyCode["Dollar"] = 36] = "Dollar";
    AltKeyCode[AltKeyCode["Ampersand"] = 38] = "Ampersand";
    AltKeyCode[AltKeyCode["Quote"] = 39] = "Quote";
    AltKeyCode[AltKeyCode["LeftParen"] = 40] = "LeftParen";
    AltKeyCode[AltKeyCode["RightParen"] = 41] = "RightParen";
    AltKeyCode[AltKeyCode["Asterisk"] = 42] = "Asterisk";
    AltKeyCode[AltKeyCode["Plus"] = 43] = "Plus";
    AltKeyCode[AltKeyCode["Comma"] = 44] = "Comma";
    AltKeyCode[AltKeyCode["Minus"] = 45] = "Minus";
    AltKeyCode[AltKeyCode["Period"] = 46] = "Period";
    AltKeyCode[AltKeyCode["Slash"] = 47] = "Slash";
    AltKeyCode[AltKeyCode["Alpha0"] = 48] = "Alpha0";
    AltKeyCode[AltKeyCode["Alpha1"] = 49] = "Alpha1";
    AltKeyCode[AltKeyCode["Alpha2"] = 50] = "Alpha2";
    AltKeyCode[AltKeyCode["Alpha3"] = 51] = "Alpha3";
    AltKeyCode[AltKeyCode["Alpha4"] = 52] = "Alpha4";
    AltKeyCode[AltKeyCode["Alpha5"] = 53] = "Alpha5";
    AltKeyCode[AltKeyCode["Alpha6"] = 54] = "Alpha6";
    AltKeyCode[AltKeyCode["Alpha7"] = 55] = "Alpha7";
    AltKeyCode[AltKeyCode["Alpha8"] = 56] = "Alpha8";
    AltKeyCode[AltKeyCode["Alpha9"] = 57] = "Alpha9";
    AltKeyCode[AltKeyCode["Colon"] = 58] = "Colon";
    AltKeyCode[AltKeyCode["Semicolon"] = 59] = "Semicolon";
    AltKeyCode[AltKeyCode["Less"] = 60] = "Less";
    AltKeyCode[AltKeyCode["Equals"] = 61] = "Equals";
    AltKeyCode[AltKeyCode["Greater"] = 62] = "Greater";
    AltKeyCode[AltKeyCode["Question"] = 63] = "Question";
    AltKeyCode[AltKeyCode["At"] = 64] = "At";
    AltKeyCode[AltKeyCode["LeftBracket"] = 91] = "LeftBracket";
    AltKeyCode[AltKeyCode["Backslash"] = 92] = "Backslash";
    AltKeyCode[AltKeyCode["RightBracket"] = 93] = "RightBracket";
    AltKeyCode[AltKeyCode["Caret"] = 94] = "Caret";
    AltKeyCode[AltKeyCode["Underscore"] = 95] = "Underscore";
    AltKeyCode[AltKeyCode["BackQuote"] = 96] = "BackQuote";
    AltKeyCode[AltKeyCode["A"] = 97] = "A";
    AltKeyCode[AltKeyCode["B"] = 98] = "B";
    AltKeyCode[AltKeyCode["C"] = 99] = "C";
    AltKeyCode[AltKeyCode["D"] = 100] = "D";
    AltKeyCode[AltKeyCode["E"] = 101] = "E";
    AltKeyCode[AltKeyCode["F"] = 102] = "F";
    AltKeyCode[AltKeyCode["G"] = 103] = "G";
    AltKeyCode[AltKeyCode["H"] = 104] = "H";
    AltKeyCode[AltKeyCode["I"] = 105] = "I";
    AltKeyCode[AltKeyCode["J"] = 106] = "J";
    AltKeyCode[AltKeyCode["K"] = 107] = "K";
    AltKeyCode[AltKeyCode["L"] = 108] = "L";
    AltKeyCode[AltKeyCode["M"] = 109] = "M";
    AltKeyCode[AltKeyCode["N"] = 110] = "N";
    AltKeyCode[AltKeyCode["O"] = 111] = "O";
    AltKeyCode[AltKeyCode["P"] = 112] = "P";
    AltKeyCode[AltKeyCode["Q"] = 113] = "Q";
    AltKeyCode[AltKeyCode["R"] = 114] = "R";
    AltKeyCode[AltKeyCode["S"] = 115] = "S";
    AltKeyCode[AltKeyCode["T"] = 116] = "T";
    AltKeyCode[AltKeyCode["U"] = 117] = "U";
    AltKeyCode[AltKeyCode["V"] = 118] = "V";
    AltKeyCode[AltKeyCode["W"] = 119] = "W";
    AltKeyCode[AltKeyCode["X"] = 120] = "X";
    AltKeyCode[AltKeyCode["Y"] = 121] = "Y";
    AltKeyCode[AltKeyCode["Z"] = 122] = "Z";
    AltKeyCode[AltKeyCode["Delete"] = 127] = "Delete";
    AltKeyCode[AltKeyCode["Keypad0"] = 256] = "Keypad0";
    AltKeyCode[AltKeyCode["Keypad1"] = 257] = "Keypad1";
    AltKeyCode[AltKeyCode["Keypad2"] = 258] = "Keypad2";
    AltKeyCode[AltKeyCode["Keypad3"] = 259] = "Keypad3";
    AltKeyCode[AltKeyCode["Keypad4"] = 260] = "Keypad4";
    AltKeyCode[AltKeyCode["Keypad5"] = 261] = "Keypad5";
    AltKeyCode[AltKeyCode["Keypad6"] = 262] = "Keypad6";
    AltKeyCode[AltKeyCode["Keypad7"] = 263] = "Keypad7";
    AltKeyCode[AltKeyCode["Keypad8"] = 264] = "Keypad8";
    AltKeyCode[AltKeyCode["Keypad9"] = 265] = "Keypad9";
    AltKeyCode[AltKeyCode["KeypadPeriod"] = 266] = "KeypadPeriod";
    AltKeyCode[AltKeyCode["KeypadDivide"] = 267] = "KeypadDivide";
    AltKeyCode[AltKeyCode["KeypadMultiply"] = 268] = "KeypadMultiply";
    AltKeyCode[AltKeyCode["KeypadMinus"] = 269] = "KeypadMinus";
    AltKeyCode[AltKeyCode["KeypadPlus"] = 270] = "KeypadPlus";
    AltKeyCode[AltKeyCode["KeypadEnter"] = 271] = "KeypadEnter";
    AltKeyCode[AltKeyCode["KeypadEquals"] = 272] = "KeypadEquals";
    AltKeyCode[AltKeyCode["UpArrow"] = 273] = "UpArrow";
    AltKeyCode[AltKeyCode["DownArrow"] = 274] = "DownArrow";
    AltKeyCode[AltKeyCode["RightArrow"] = 275] = "RightArrow";
    AltKeyCode[AltKeyCode["LeftArrow"] = 276] = "LeftArrow";
    AltKeyCode[AltKeyCode["Insert"] = 277] = "Insert";
    AltKeyCode[AltKeyCode["Home"] = 278] = "Home";
    AltKeyCode[AltKeyCode["End"] = 279] = "End";
    AltKeyCode[AltKeyCode["PageUp"] = 280] = "PageUp";
    AltKeyCode[AltKeyCode["PageDown"] = 281] = "PageDown";
    AltKeyCode[AltKeyCode["F1"] = 282] = "F1";
    AltKeyCode[AltKeyCode["F2"] = 283] = "F2";
    AltKeyCode[AltKeyCode["F3"] = 284] = "F3";
    AltKeyCode[AltKeyCode["F4"] = 285] = "F4";
    AltKeyCode[AltKeyCode["F5"] = 286] = "F5";
    AltKeyCode[AltKeyCode["F6"] = 287] = "F6";
    AltKeyCode[AltKeyCode["F7"] = 288] = "F7";
    AltKeyCode[AltKeyCode["F8"] = 289] = "F8";
    AltKeyCode[AltKeyCode["F9"] = 290] = "F9";
    AltKeyCode[AltKeyCode["F10"] = 291] = "F10";
    AltKeyCode[AltKeyCode["F11"] = 292] = "F11";
    AltKeyCode[AltKeyCode["F12"] = 293] = "F12";
    AltKeyCode[AltKeyCode["F13"] = 294] = "F13";
    AltKeyCode[AltKeyCode["F14"] = 295] = "F14";
    AltKeyCode[AltKeyCode["F15"] = 296] = "F15";
    AltKeyCode[AltKeyCode["Numlock"] = 300] = "Numlock";
    AltKeyCode[AltKeyCode["CapsLock"] = 301] = "CapsLock";
    AltKeyCode[AltKeyCode["ScrollLock"] = 302] = "ScrollLock";
    AltKeyCode[AltKeyCode["RightShift"] = 303] = "RightShift";
    AltKeyCode[AltKeyCode["LeftShift"] = 304] = "LeftShift";
    AltKeyCode[AltKeyCode["RightControl"] = 305] = "RightControl";
    AltKeyCode[AltKeyCode["LeftControl"] = 306] = "LeftControl";
    AltKeyCode[AltKeyCode["RightAlt"] = 307] = "RightAlt";
    AltKeyCode[AltKeyCode["LeftAlt"] = 308] = "LeftAlt";
    AltKeyCode[AltKeyCode["RightCommand"] = 309] = "RightCommand";
    AltKeyCode[AltKeyCode["RightApple"] = 309] = "RightApple";
    AltKeyCode[AltKeyCode["LeftCommand"] = 310] = "LeftCommand";
    AltKeyCode[AltKeyCode["LeftApple"] = 310] = "LeftApple";
    AltKeyCode[AltKeyCode["LeftWindows"] = 311] = "LeftWindows";
    AltKeyCode[AltKeyCode["RightWindows"] = 312] = "RightWindows";
    AltKeyCode[AltKeyCode["AltGr"] = 313] = "AltGr";
    AltKeyCode[AltKeyCode["Help"] = 315] = "Help";
    AltKeyCode[AltKeyCode["Print"] = 316] = "Print";
    AltKeyCode[AltKeyCode["SysReq"] = 317] = "SysReq";
    AltKeyCode[AltKeyCode["Break"] = 318] = "Break";
    AltKeyCode[AltKeyCode["Menu"] = 319] = "Menu";
    AltKeyCode[AltKeyCode["Mouse0"] = 323] = "Mouse0";
    AltKeyCode[AltKeyCode["Mouse1"] = 324] = "Mouse1";
    AltKeyCode[AltKeyCode["Mouse2"] = 325] = "Mouse2";
    AltKeyCode[AltKeyCode["Mouse3"] = 326] = "Mouse3";
    AltKeyCode[AltKeyCode["Mouse4"] = 327] = "Mouse4";
    AltKeyCode[AltKeyCode["Mouse5"] = 328] = "Mouse5";
    AltKeyCode[AltKeyCode["Mouse6"] = 329] = "Mouse6";
    AltKeyCode[AltKeyCode["JoystickButton0"] = 330] = "JoystickButton0";
    AltKeyCode[AltKeyCode["JoystickButton1"] = 331] = "JoystickButton1";
    AltKeyCode[AltKeyCode["JoystickButton2"] = 332] = "JoystickButton2";
    AltKeyCode[AltKeyCode["JoystickButton3"] = 333] = "JoystickButton3";
    AltKeyCode[AltKeyCode["JoystickButton4"] = 334] = "JoystickButton4";
    AltKeyCode[AltKeyCode["JoystickButton5"] = 335] = "JoystickButton5";
    AltKeyCode[AltKeyCode["JoystickButton6"] = 336] = "JoystickButton6";
    AltKeyCode[AltKeyCode["JoystickButton7"] = 337] = "JoystickButton7";
    AltKeyCode[AltKeyCode["JoystickButton8"] = 338] = "JoystickButton8";
    AltKeyCode[AltKeyCode["JoystickButton9"] = 339] = "JoystickButton9";
    AltKeyCode[AltKeyCode["JoystickButton10"] = 340] = "JoystickButton10";
    AltKeyCode[AltKeyCode["JoystickButton11"] = 341] = "JoystickButton11";
    AltKeyCode[AltKeyCode["JoystickButton12"] = 342] = "JoystickButton12";
    AltKeyCode[AltKeyCode["JoystickButton13"] = 343] = "JoystickButton13";
    AltKeyCode[AltKeyCode["JoystickButton14"] = 344] = "JoystickButton14";
    AltKeyCode[AltKeyCode["JoystickButton15"] = 345] = "JoystickButton15";
    AltKeyCode[AltKeyCode["JoystickButton16"] = 346] = "JoystickButton16";
    AltKeyCode[AltKeyCode["JoystickButton17"] = 347] = "JoystickButton17";
    AltKeyCode[AltKeyCode["JoystickButton18"] = 348] = "JoystickButton18";
    AltKeyCode[AltKeyCode["JoystickButton19"] = 349] = "JoystickButton19";
    AltKeyCode[AltKeyCode["Joystick1Button0"] = 350] = "Joystick1Button0";
    AltKeyCode[AltKeyCode["Joystick1Button1"] = 351] = "Joystick1Button1";
    AltKeyCode[AltKeyCode["Joystick1Button2"] = 352] = "Joystick1Button2";
    AltKeyCode[AltKeyCode["Joystick1Button3"] = 353] = "Joystick1Button3";
    AltKeyCode[AltKeyCode["Joystick1Button4"] = 354] = "Joystick1Button4";
    AltKeyCode[AltKeyCode["Joystick1Button5"] = 355] = "Joystick1Button5";
    AltKeyCode[AltKeyCode["Joystick1Button6"] = 356] = "Joystick1Button6";
    AltKeyCode[AltKeyCode["Joystick1Button7"] = 357] = "Joystick1Button7";
    AltKeyCode[AltKeyCode["Joystick1Button8"] = 358] = "Joystick1Button8";
    AltKeyCode[AltKeyCode["Joystick1Button9"] = 359] = "Joystick1Button9";
    AltKeyCode[AltKeyCode["Joystick1Button10"] = 360] = "Joystick1Button10";
    AltKeyCode[AltKeyCode["Joystick1Button11"] = 361] = "Joystick1Button11";
    AltKeyCode[AltKeyCode["Joystick1Button12"] = 362] = "Joystick1Button12";
    AltKeyCode[AltKeyCode["Joystick1Button13"] = 363] = "Joystick1Button13";
    AltKeyCode[AltKeyCode["Joystick1Button14"] = 364] = "Joystick1Button14";
    AltKeyCode[AltKeyCode["Joystick1Button15"] = 365] = "Joystick1Button15";
    AltKeyCode[AltKeyCode["Joystick1Button16"] = 366] = "Joystick1Button16";
    AltKeyCode[AltKeyCode["Joystick1Button17"] = 367] = "Joystick1Button17";
    AltKeyCode[AltKeyCode["Joystick1Button18"] = 368] = "Joystick1Button18";
    AltKeyCode[AltKeyCode["Joystick1Button19"] = 369] = "Joystick1Button19";
    AltKeyCode[AltKeyCode["Joystick2Button0"] = 370] = "Joystick2Button0";
    AltKeyCode[AltKeyCode["Joystick2Button1"] = 371] = "Joystick2Button1";
    AltKeyCode[AltKeyCode["Joystick2Button2"] = 372] = "Joystick2Button2";
    AltKeyCode[AltKeyCode["Joystick2Button3"] = 373] = "Joystick2Button3";
    AltKeyCode[AltKeyCode["Joystick2Button4"] = 374] = "Joystick2Button4";
    AltKeyCode[AltKeyCode["Joystick2Button5"] = 375] = "Joystick2Button5";
    AltKeyCode[AltKeyCode["Joystick2Button6"] = 376] = "Joystick2Button6";
    AltKeyCode[AltKeyCode["Joystick2Button7"] = 377] = "Joystick2Button7";
    AltKeyCode[AltKeyCode["Joystick2Button8"] = 378] = "Joystick2Button8";
    AltKeyCode[AltKeyCode["Joystick2Button9"] = 379] = "Joystick2Button9";
    AltKeyCode[AltKeyCode["Joystick2Button10"] = 380] = "Joystick2Button10";
    AltKeyCode[AltKeyCode["Joystick2Button11"] = 381] = "Joystick2Button11";
    AltKeyCode[AltKeyCode["Joystick2Button12"] = 382] = "Joystick2Button12";
    AltKeyCode[AltKeyCode["Joystick2Button13"] = 383] = "Joystick2Button13";
    AltKeyCode[AltKeyCode["Joystick2Button14"] = 384] = "Joystick2Button14";
    AltKeyCode[AltKeyCode["Joystick2Button15"] = 385] = "Joystick2Button15";
    AltKeyCode[AltKeyCode["Joystick2Button16"] = 386] = "Joystick2Button16";
    AltKeyCode[AltKeyCode["Joystick2Button17"] = 387] = "Joystick2Button17";
    AltKeyCode[AltKeyCode["Joystick2Button18"] = 388] = "Joystick2Button18";
    AltKeyCode[AltKeyCode["Joystick2Button19"] = 389] = "Joystick2Button19";
    AltKeyCode[AltKeyCode["Joystick3Button0"] = 390] = "Joystick3Button0";
    AltKeyCode[AltKeyCode["Joystick3Button1"] = 391] = "Joystick3Button1";
    AltKeyCode[AltKeyCode["Joystick3Button2"] = 392] = "Joystick3Button2";
    AltKeyCode[AltKeyCode["Joystick3Button3"] = 393] = "Joystick3Button3";
    AltKeyCode[AltKeyCode["Joystick3Button4"] = 394] = "Joystick3Button4";
    AltKeyCode[AltKeyCode["Joystick3Button5"] = 395] = "Joystick3Button5";
    AltKeyCode[AltKeyCode["Joystick3Button6"] = 396] = "Joystick3Button6";
    AltKeyCode[AltKeyCode["Joystick3Button7"] = 397] = "Joystick3Button7";
    AltKeyCode[AltKeyCode["Joystick3Button8"] = 398] = "Joystick3Button8";
    AltKeyCode[AltKeyCode["Joystick3Button9"] = 399] = "Joystick3Button9";
    AltKeyCode[AltKeyCode["Joystick3Button10"] = 400] = "Joystick3Button10";
    AltKeyCode[AltKeyCode["Joystick3Button11"] = 401] = "Joystick3Button11";
    AltKeyCode[AltKeyCode["Joystick3Button12"] = 402] = "Joystick3Button12";
    AltKeyCode[AltKeyCode["Joystick3Button13"] = 403] = "Joystick3Button13";
    AltKeyCode[AltKeyCode["Joystick3Button14"] = 404] = "Joystick3Button14";
    AltKeyCode[AltKeyCode["Joystick3Button15"] = 405] = "Joystick3Button15";
    AltKeyCode[AltKeyCode["Joystick3Button16"] = 406] = "Joystick3Button16";
    AltKeyCode[AltKeyCode["Joystick3Button17"] = 407] = "Joystick3Button17";
    AltKeyCode[AltKeyCode["Joystick3Button18"] = 408] = "Joystick3Button18";
    AltKeyCode[AltKeyCode["Joystick3Button19"] = 409] = "Joystick3Button19";
    AltKeyCode[AltKeyCode["Joystick4Button0"] = 410] = "Joystick4Button0";
    AltKeyCode[AltKeyCode["Joystick4Button1"] = 411] = "Joystick4Button1";
    AltKeyCode[AltKeyCode["Joystick4Button2"] = 412] = "Joystick4Button2";
    AltKeyCode[AltKeyCode["Joystick4Button3"] = 413] = "Joystick4Button3";
    AltKeyCode[AltKeyCode["Joystick4Button4"] = 414] = "Joystick4Button4";
    AltKeyCode[AltKeyCode["Joystick4Button5"] = 415] = "Joystick4Button5";
    AltKeyCode[AltKeyCode["Joystick4Button6"] = 416] = "Joystick4Button6";
    AltKeyCode[AltKeyCode["Joystick4Button7"] = 417] = "Joystick4Button7";
    AltKeyCode[AltKeyCode["Joystick4Button8"] = 418] = "Joystick4Button8";
    AltKeyCode[AltKeyCode["Joystick4Button9"] = 419] = "Joystick4Button9";
    AltKeyCode[AltKeyCode["Joystick4Button10"] = 420] = "Joystick4Button10";
    AltKeyCode[AltKeyCode["Joystick4Button11"] = 421] = "Joystick4Button11";
    AltKeyCode[AltKeyCode["Joystick4Button12"] = 422] = "Joystick4Button12";
    AltKeyCode[AltKeyCode["Joystick4Button13"] = 423] = "Joystick4Button13";
    AltKeyCode[AltKeyCode["Joystick4Button14"] = 424] = "Joystick4Button14";
    AltKeyCode[AltKeyCode["Joystick4Button15"] = 425] = "Joystick4Button15";
    AltKeyCode[AltKeyCode["Joystick4Button16"] = 426] = "Joystick4Button16";
    AltKeyCode[AltKeyCode["Joystick4Button17"] = 427] = "Joystick4Button17";
    AltKeyCode[AltKeyCode["Joystick4Button18"] = 428] = "Joystick4Button18";
    AltKeyCode[AltKeyCode["Joystick4Button19"] = 429] = "Joystick4Button19";
    AltKeyCode[AltKeyCode["Joystick5Button0"] = 430] = "Joystick5Button0";
    AltKeyCode[AltKeyCode["Joystick5Button1"] = 431] = "Joystick5Button1";
    AltKeyCode[AltKeyCode["Joystick5Button2"] = 432] = "Joystick5Button2";
    AltKeyCode[AltKeyCode["Joystick5Button3"] = 433] = "Joystick5Button3";
    AltKeyCode[AltKeyCode["Joystick5Button4"] = 434] = "Joystick5Button4";
    AltKeyCode[AltKeyCode["Joystick5Button5"] = 435] = "Joystick5Button5";
    AltKeyCode[AltKeyCode["Joystick5Button6"] = 436] = "Joystick5Button6";
    AltKeyCode[AltKeyCode["Joystick5Button7"] = 437] = "Joystick5Button7";
    AltKeyCode[AltKeyCode["Joystick5Button8"] = 438] = "Joystick5Button8";
    AltKeyCode[AltKeyCode["Joystick5Button9"] = 439] = "Joystick5Button9";
    AltKeyCode[AltKeyCode["Joystick5Button10"] = 440] = "Joystick5Button10";
    AltKeyCode[AltKeyCode["Joystick5Button11"] = 441] = "Joystick5Button11";
    AltKeyCode[AltKeyCode["Joystick5Button12"] = 442] = "Joystick5Button12";
    AltKeyCode[AltKeyCode["Joystick5Button13"] = 443] = "Joystick5Button13";
    AltKeyCode[AltKeyCode["Joystick5Button14"] = 444] = "Joystick5Button14";
    AltKeyCode[AltKeyCode["Joystick5Button15"] = 445] = "Joystick5Button15";
    AltKeyCode[AltKeyCode["Joystick5Button16"] = 446] = "Joystick5Button16";
    AltKeyCode[AltKeyCode["Joystick5Button17"] = 447] = "Joystick5Button17";
    AltKeyCode[AltKeyCode["Joystick5Button18"] = 448] = "Joystick5Button18";
    AltKeyCode[AltKeyCode["Joystick5Button19"] = 449] = "Joystick5Button19";
    AltKeyCode[AltKeyCode["Joystick6Button0"] = 450] = "Joystick6Button0";
    AltKeyCode[AltKeyCode["Joystick6Button1"] = 451] = "Joystick6Button1";
    AltKeyCode[AltKeyCode["Joystick6Button2"] = 452] = "Joystick6Button2";
    AltKeyCode[AltKeyCode["Joystick6Button3"] = 453] = "Joystick6Button3";
    AltKeyCode[AltKeyCode["Joystick6Button4"] = 454] = "Joystick6Button4";
    AltKeyCode[AltKeyCode["Joystick6Button5"] = 455] = "Joystick6Button5";
    AltKeyCode[AltKeyCode["Joystick6Button6"] = 456] = "Joystick6Button6";
    AltKeyCode[AltKeyCode["Joystick6Button7"] = 457] = "Joystick6Button7";
    AltKeyCode[AltKeyCode["Joystick6Button8"] = 458] = "Joystick6Button8";
    AltKeyCode[AltKeyCode["Joystick6Button9"] = 459] = "Joystick6Button9";
    AltKeyCode[AltKeyCode["Joystick6Button10"] = 460] = "Joystick6Button10";
    AltKeyCode[AltKeyCode["Joystick6Button11"] = 461] = "Joystick6Button11";
    AltKeyCode[AltKeyCode["Joystick6Button12"] = 462] = "Joystick6Button12";
    AltKeyCode[AltKeyCode["Joystick6Button13"] = 463] = "Joystick6Button13";
    AltKeyCode[AltKeyCode["Joystick6Button14"] = 464] = "Joystick6Button14";
    AltKeyCode[AltKeyCode["Joystick6Button15"] = 465] = "Joystick6Button15";
    AltKeyCode[AltKeyCode["Joystick6Button16"] = 466] = "Joystick6Button16";
    AltKeyCode[AltKeyCode["Joystick6Button17"] = 467] = "Joystick6Button17";
    AltKeyCode[AltKeyCode["Joystick6Button18"] = 468] = "Joystick6Button18";
    AltKeyCode[AltKeyCode["Joystick6Button19"] = 469] = "Joystick6Button19";
    AltKeyCode[AltKeyCode["Joystick7Button0"] = 470] = "Joystick7Button0";
    AltKeyCode[AltKeyCode["Joystick7Button1"] = 471] = "Joystick7Button1";
    AltKeyCode[AltKeyCode["Joystick7Button2"] = 472] = "Joystick7Button2";
    AltKeyCode[AltKeyCode["Joystick7Button3"] = 473] = "Joystick7Button3";
    AltKeyCode[AltKeyCode["Joystick7Button4"] = 474] = "Joystick7Button4";
    AltKeyCode[AltKeyCode["Joystick7Button5"] = 475] = "Joystick7Button5";
    AltKeyCode[AltKeyCode["Joystick7Button6"] = 476] = "Joystick7Button6";
    AltKeyCode[AltKeyCode["Joystick7Button7"] = 477] = "Joystick7Button7";
    AltKeyCode[AltKeyCode["Joystick7Button8"] = 478] = "Joystick7Button8";
    AltKeyCode[AltKeyCode["Joystick7Button9"] = 479] = "Joystick7Button9";
    AltKeyCode[AltKeyCode["Joystick7Button10"] = 480] = "Joystick7Button10";
    AltKeyCode[AltKeyCode["Joystick7Button11"] = 481] = "Joystick7Button11";
    AltKeyCode[AltKeyCode["Joystick7Button12"] = 482] = "Joystick7Button12";
    AltKeyCode[AltKeyCode["Joystick7Button13"] = 483] = "Joystick7Button13";
    AltKeyCode[AltKeyCode["Joystick7Button14"] = 484] = "Joystick7Button14";
    AltKeyCode[AltKeyCode["Joystick7Button15"] = 485] = "Joystick7Button15";
    AltKeyCode[AltKeyCode["Joystick7Button16"] = 486] = "Joystick7Button16";
    AltKeyCode[AltKeyCode["Joystick7Button17"] = 487] = "Joystick7Button17";
    AltKeyCode[AltKeyCode["Joystick7Button18"] = 488] = "Joystick7Button18";
    AltKeyCode[AltKeyCode["Joystick7Button19"] = 489] = "Joystick7Button19";
    AltKeyCode[AltKeyCode["Joystick8Button0"] = 490] = "Joystick8Button0";
    AltKeyCode[AltKeyCode["Joystick8Button1"] = 491] = "Joystick8Button1";
    AltKeyCode[AltKeyCode["Joystick8Button2"] = 492] = "Joystick8Button2";
    AltKeyCode[AltKeyCode["Joystick8Button3"] = 493] = "Joystick8Button3";
    AltKeyCode[AltKeyCode["Joystick8Button4"] = 494] = "Joystick8Button4";
    AltKeyCode[AltKeyCode["Joystick8Button5"] = 495] = "Joystick8Button5";
    AltKeyCode[AltKeyCode["Joystick8Button6"] = 496] = "Joystick8Button6";
    AltKeyCode[AltKeyCode["Joystick8Button7"] = 497] = "Joystick8Button7";
    AltKeyCode[AltKeyCode["Joystick8Button8"] = 498] = "Joystick8Button8";
    AltKeyCode[AltKeyCode["Joystick8Button9"] = 499] = "Joystick8Button9";
    AltKeyCode[AltKeyCode["Joystick8Button10"] = 500] = "Joystick8Button10";
    AltKeyCode[AltKeyCode["Joystick8Button11"] = 501] = "Joystick8Button11";
    AltKeyCode[AltKeyCode["Joystick8Button12"] = 502] = "Joystick8Button12";
    AltKeyCode[AltKeyCode["Joystick8Button13"] = 503] = "Joystick8Button13";
    AltKeyCode[AltKeyCode["Joystick8Button14"] = 504] = "Joystick8Button14";
    AltKeyCode[AltKeyCode["Joystick8Button15"] = 505] = "Joystick8Button15";
    AltKeyCode[AltKeyCode["Joystick8Button16"] = 506] = "Joystick8Button16";
    AltKeyCode[AltKeyCode["Joystick8Button17"] = 507] = "Joystick8Button17";
    AltKeyCode[AltKeyCode["Joystick8Button18"] = 508] = "Joystick8Button18";
    AltKeyCode[AltKeyCode["Joystick8Button19"] = 509] = "Joystick8Button19";
})(AltKeyCode = exports.AltKeyCode || (exports.AltKeyCode = {}));
