// appium-altunity-plugin/build/src/commands/element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProperty = exports.getAttribute = exports.getText = exports.getElementRect = exports.getLocation = exports.elementDisplayed = exports.click = exports.unityElementGuard = exports.UnityElement = exports.ALL_COMPONENTS_PROP = exports.UNITY_ELEMENT_PREFIX = void 0;
const support_1 = require("appium/support");
const driver_1 = require("appium/driver");
const alt_element_1 = require("../client/alt-element");
exports.UNITY_ELEMENT_PREFIX = 'unity-element-';
exports.ALL_COMPONENTS_PROP = '*';
const COMPONENT_PROPERTY_RE = /(.+):(.+)/;
class UnityElement {
    constructor(element) {
        this.element = element;
        this.id = `${exports.UNITY_ELEMENT_PREFIX}${element.id}`;
    }
    get location() {
        return { x: this.element.x, y: this.element.y };
    }
    get asW3CElementObject() {
        return { [support_1.util.W3C_WEB_ELEMENT_IDENTIFIER]: this.id };
    }
}
exports.UnityElement = UnityElement;
async function unityElementGuard(next, elId, fn) {
    return await this.unityContextGuard(next, async () => {
        const el = this.unityElements[elId];
        if (!el) {
            throw new driver_1.errors.NoSuchElementError(undefined);
        }
        return await fn(el);
    });
}
exports.unityElementGuard = unityElementGuard;
async function click(next, driver, elId) {
    return await this.unityElementGuard(next, elId, async (el) => await el.element.tapViaCoords());
}
exports.click = click;
async function elementDisplayed(next, driver, elId) {
    return await this.unityElementGuard(next, elId, async (el) => {
        const { width, height } = await this.getCachedScreenDims(driver);
        const { x, y } = el.element.asUnityObject();
        return x >= 0 && x <= width && y >= 0 && y <= height;
    });
}
exports.elementDisplayed = elementDisplayed;
async function getLocation(next, _, elId) {
    return await this.unityElementGuard(next, elId, async (el) => el.location);
}
exports.getLocation = getLocation;
async function getElementRect(next, _, elId) {
    // just report 0 for width and height since we don't know them
    return await this.unityElementGuard(next, elId, async (el) => ({ ...el.location, width: 0, height: 0 }));
}
exports.getElementRect = getElementRect;
async function getText(next, _, elId) {
    return await this.unityElementGuard(next, elId, async (el) => {
        try {
            return await el.element.getText();
        }
        catch (err) {
            if (err.message.includes('Component TMP_InputField')) {
                return null;
            }
            throw err;
        }
    });
}
exports.getText = getText;
async function getAttribute(next, _, attr, elId) {
    return await this.unityElementGuard(next, elId, async (el) => {
        if (!alt_element_1.ALT_ELEMENT_KEYS.includes(attr)) {
            throw new Error(`Cannot get '${attr}' attribute from element`);
        }
        const data = el.element.asUnityObject();
        return data[attr];
    });
}
exports.getAttribute = getAttribute;
async function getProperty(next, _, prop, elId) {
    return await this.unityElementGuard(next, elId, async (el) => {
        // then check if it's of the form 'component:property' to return a component prop directly
        const componentMatch = COMPONENT_PROPERTY_RE.exec(prop);
        if (componentMatch) {
            const [, component, property] = componentMatch;
            return JSON.stringify(await el.element.getComponentProperty(component, property, ''));
        }
        const allComponents = await el.element.getComponents();
        // otherwise if it's the magic '*', return all components
        if (prop === exports.ALL_COMPONENTS_PROP) {
            return JSON.stringify(allComponents.map((c) => c.toObject()));
        }
        // otherwise it must be the name of a component to return
        const components = allComponents.filter((c) => c.name === prop);
        if (!components[0]) {
            throw new Error(`Element had no component named '${prop}'`);
        }
        return JSON.stringify(components[0].toObject());
    });
}
exports.getProperty = getProperty;
