// appium-dashboard/lib/models/build.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Build = void 0;
const sequelize_typescript_1 = require("sequelize-typescript");
const sequelize_1 = require("sequelize");
const _1 = require(".");
let Build = class Build extends sequelize_typescript_1.Model {
};
__decorate([
    (0, sequelize_typescript_1.Column)({
        primaryKey: true,
        autoIncrement: true,
        type: sequelize_1.DataTypes.INTEGER,
    })
], Build.prototype, "id", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.STRING,
        unique: true,
    })
], Build.prototype, "build_id", void 0);
__decorate([
    (0, sequelize_typescript_1.Column)({
        allowNull: true,
        type: sequelize_1.DataTypes.INTEGER,
    }),
    (0, sequelize_typescript_1.ForeignKey)(() => _1.Project)
], Build.prototype, "project_id", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.STRING,
    })
], Build.prototype, "name", void 0);
__decorate([
    (0, sequelize_typescript_1.HasMany)(() => _1.Session, { sourceKey: "build_id" })
], Build.prototype, "sessions", void 0);
__decorate([
    (0, sequelize_typescript_1.BelongsTo)(() => _1.Project)
], Build.prototype, "project", void 0);
Build = __decorate([
    (0, sequelize_typescript_1.Table)({
        tableName: "builds",
        timestamps: true,
        underscored: true,
        createdAt: "created_at",
        updatedAt: "updated_at",
    })
], Build);
exports.Build = Build;
