// appium-dashboard/lib/models/http-logs.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpLogs = void 0;
const sequelize_typescript_1 = require("sequelize-typescript");
const sequelize_1 = require("sequelize");
const session_1 = require("./session");
const utils_1 = require("../utils");
let HttpLogs = class HttpLogs extends sequelize_typescript_1.Model {
};
__decorate([
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true,
    })
], HttpLogs.prototype, "id", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.STRING,
    }),
    (0, sequelize_typescript_1.ForeignKey)(() => session_1.Session)
], HttpLogs.prototype, "session_id", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.STRING,
    })
], HttpLogs.prototype, "url", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "method", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)(Object.assign({ type: sequelize_1.DataTypes.TEXT }, (0, utils_1.customModelColumn)({ name: "request_headers", json: true })))
], HttpLogs.prototype, "request_headers", void 0);
__decorate([
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "request_post_data", void 0);
__decorate([
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.STRING,
    })
], HttpLogs.prototype, "request_content_type", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "request_type", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "context", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.INTEGER,
    })
], HttpLogs.prototype, "response_status", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "response_status_text", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(false),
    (0, sequelize_typescript_1.Column)(Object.assign({ type: sequelize_1.DataTypes.TEXT }, (0, utils_1.customModelColumn)({ name: "response_headers", json: true })))
], HttpLogs.prototype, "response_headers", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "response_content_type", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "response_body", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.TEXT,
    })
], HttpLogs.prototype, "remote_ip_address", void 0);
__decorate([
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.DATE,
    })
], HttpLogs.prototype, "start_time", void 0);
__decorate([
    (0, sequelize_typescript_1.AllowNull)(true),
    (0, sequelize_typescript_1.Column)({
        type: sequelize_1.DataTypes.DATE,
    })
], HttpLogs.prototype, "end_time", void 0);
__decorate([
    (0, sequelize_typescript_1.BelongsTo)(() => session_1.Session, { foreignKey: "session_id" })
], HttpLogs.prototype, "session", void 0);
HttpLogs = __decorate([
    (0, sequelize_typescript_1.Table)({
        tableName: "http_logs",
        timestamps: true,
        underscored: true,
        createdAt: "created_at",
        updatedAt: "updated_at",
    })
], HttpLogs);
exports.HttpLogs = HttpLogs;
