// appium-dashboard/lib/models/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpLogs = exports.Profiling = exports.Project = exports.Build = exports.CommandLogs = exports.Logs = exports.Session = void 0;
var session_1 = require("./session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return session_1.Session; } });
var logs_1 = require("./logs");
Object.defineProperty(exports, "Logs", { enumerable: true, get: function () { return logs_1.Logs; } });
var command_logs_1 = require("./command-logs");
Object.defineProperty(exports, "CommandLogs", { enumerable: true, get: function () { return command_logs_1.CommandLogs; } });
var build_1 = require("./build");
Object.defineProperty(exports, "Build", { enumerable: true, get: function () { return build_1.Build; } });
var project_1 = require("./project");
Object.defineProperty(exports, "Project", { enumerable: true, get: function () { return project_1.Project; } });
var profiling_1 = require("./profiling");
Object.defineProperty(exports, "Profiling", { enumerable: true, get: function () { return profiling_1.Profiling; } });
var http_logs_1 = require("./http-logs");
Object.defineProperty(exports, "HttpLogs", { enumerable: true, get: function () { return http_logs_1.HttpLogs; } });
