// appium-gestures-plugin/lib/gestures/doubleTap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Element = __importStar(require("../element"));
async function doubleTap(elementId, driver) {
    {
        const value = await driver.getElementRect(elementId);
        const { x, y } = await Element.getCenter(value);
        const androidPauseAction = {
            duration: 0,
            type: 'pause',
        };
        const actionsData = [
            {
                id: 'finger1',
                type: 'pointer',
                parameters: { pointerType: 'touch' },
                actions: [
                    { duration: 0, x, y, type: 'pointerMove', origin: 'viewport' },
                    { button: 0, type: 'pointerDown' },
                    { duration: 200, type: 'pause' },
                    { button: 0, type: 'pointerUp' },
                    { duration: 40, type: 'pause' },
                    { button: 0, type: 'pointerDown' },
                    { duration: 200, type: 'pause' },
                    { button: 0, type: 'pointerUp' },
                ],
            },
        ];
        if (driver.caps.automationName !== 'XCuiTest') {
            actionsData[0].actions.unshift(androidPauseAction);
        }
        await driver.performActions(actionsData);
    }
}
exports.default = doubleTap;
