// appium-gestures-plugin/lib/gestures/dragAndDrop.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Element = __importStar(require("../element"));
const logger_1 = __importDefault(require("../logger"));
async function dragAndDrop(sourceId, destinationId, driver) {
    const [source, destination] = await Promise.all([
        driver.getElementRect(sourceId),
        driver.getElementRect(destinationId),
    ]);
    const [{ x: sourceX, y: sourceY }, { x: destinationX, y: destinationY }] = await Promise.all([
        Element.getCenter(source),
        Element.getCenter(destination),
    ]);
    const androidPauseAction = {
        duration: 0,
        type: 'pause',
    };
    const actionsData = [
        {
            id: 'finger',
            type: 'pointer',
            parameters: { pointerType: 'touch' },
            actions: [
                { duration: 0, x: sourceX, y: sourceY, type: 'pointerMove', origin: 'viewport' },
                { button: 1, type: 'pointerDown' },
                { duration: 600, type: 'pause' },
                {
                    duration: 600,
                    x: destinationX,
                    y: destinationY,
                    type: 'pointerMove',
                    origin: 'viewport',
                },
                { button: 1, type: 'pointerUp' },
            ],
        },
    ];
    if (driver.caps.automationName === 'XCuiTest') {
        await driver.performActions(actionsData);
    }
    else {
        logger_1.default.info('Drag and Drop for android');
        const androidActions = actionsData;
        androidActions[0].actions.unshift(androidPauseAction);
        await driver.performActions(actionsData);
    }
}
exports.default = dragAndDrop;
