// appium-gestures-plugin/lib/gestures/swipe.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scrollElementIntoView = exports.swipe = void 0;
const Element = __importStar(require("../element"));
const logger_1 = __importDefault(require("../logger"));
const MAGIC_NUMBER = 49;
async function swipe(elementId, percentage, direction, driver) {
    const swipeActions = [];
    const value = await driver.getElementRect(elementId);
    logger_1.default.info(`Swiping ${direction} at ${percentage}% of the element ${elementId}`);
    const pointer = getDirectionActions(direction, value, percentage);
    const actionsData = getActionsData(elementId, pointer, driver);
    swipeActions.push(actionsData);
    await driver.performActions(swipeActions);
}
exports.swipe = swipe;
async function scrollElementIntoView(config) {
    const { scrollableView, strategy, selector, percentage, direction, maxCount = 5, driver, } = config;
    for (let count = 0; count < maxCount && !(await isElementFound(driver, strategy, selector)); count++) {
        logger_1.default.info('Swiping now...');
        await swipe(scrollableView, percentage, direction, driver);
    }
}
exports.scrollElementIntoView = scrollElementIntoView;
function getDirectionActions(direction, value, percentage) {
    const { x, y } = Element.getCenter(value);
    const directionActions = {
        left: {
            sourceX: x + (value.width * MAGIC_NUMBER) / 100,
            sourceY: y,
            destinationX: (x + (value.width * MAGIC_NUMBER) / 100) * (percentage / 100),
            destinationY: y,
        },
        right: {
            sourceX: (x + (value.width * MAGIC_NUMBER) / 100) * (percentage / 100),
            sourceY: y,
            destinationX: x + (value.width * MAGIC_NUMBER) / 100,
            destinationY: y,
        },
        up: {
            sourceX: x,
            sourceY: y + (value.height * MAGIC_NUMBER) / 100,
            destinationX: x,
            destinationY: y - percentage / 100,
        },
        down: {
            sourceX: x,
            sourceY: y - percentage / 100,
            destinationX: x,
            destinationY: y + (value.height * MAGIC_NUMBER) / 100,
        },
    };
    return directionActions[direction];
}
function getActionsData(elementId, pointer, driver) {
    const actionsData = {
        id: `${elementId}`,
        type: 'pointer',
        parameters: { pointerType: 'touch' },
        actions: [
            {
                duration: 0,
                x: pointer.sourceX,
                y: pointer.sourceY,
                type: 'pointerMove',
                origin: 'viewport',
            },
            { button: 1, type: 'pointerDown' },
            { duration: 600, type: 'pause' },
            {
                duration: 600,
                x: pointer.destinationX,
                y: pointer.destinationY,
                type: 'pointerMove',
                origin: 'viewport',
            },
            { button: 1, type: 'pointerUp' },
        ],
    };
    if (driver.caps.automationName !== 'XCuiTest') {
        actionsData.actions.unshift({ duration: 0, type: 'pause' });
    }
    return actionsData;
}
async function isElementFound(driver, strategy, selector) {
    try {
        logger_1.default.info(`Checking if ${strategy} element '${selector}' is present`);
        await driver.findElement(strategy, selector);
        logger_1.default.info(`Element '${selector}' is found`);
        return true;
    }
    catch (e) {
        logger_1.default.info(`Element '${selector}' is not found`);
        return false;
    }
}
