// appium-ocr-plugin/build/test/e2e/plugin-e2e-specs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("../..");
const webdriverio_1 = require("webdriverio");
const webdriver_1 = require("webdriver");
const earl_1 = require("earl");
const TEST_APP_PATH = process.env.TEST_APP_PATH;
const TEST_PLATFORM = process.env.TEST_PLATFORM || 'iOS';
if (!TEST_APP_PATH) {
    throw new Error(`Must define TEST_APP_PATH`);
}
let capabilities = {
    'appium:noReset': true,
    'appium:app': TEST_APP_PATH,
};
if (TEST_PLATFORM === 'iOS') {
    capabilities = Object.assign(Object.assign({}, capabilities), { platformName: 'iOS', 'appium:automationName': 'XCUITest', 'appium:platformVersion': '16.2', 'appium:deviceName': 'iPhone 14' });
}
else {
    capabilities = Object.assign(Object.assign({}, capabilities), { platformName: 'Android', 'appium:automationName': 'UiAutomator2', 'appium:deviceName': 'Android', 'appium:settings[ocrDownsampleFactor]': 2.43 });
}
const WDIO_PARAMS = {
    hostname: '127.0.0.1',
    port: 4723,
    path: '/',
    connectionRetryCount: 0,
    logLevel: 'silent',
    capabilities
};
function updateWdioBrowser(browser) {
    browser.addCommand('getOcrText', (0, webdriver_1.command)('POST', '/session/:sessionId/appium/ocr', {
        command: 'getOcrText',
        description: 'Get all OCR text',
        ref: '',
        variables: [],
        parameters: []
    }));
}
describe('AppiumOcrPlugin', function () {
    let driver;
    before(function () {
        return __awaiter(this, void 0, void 0, function* () {
            driver = (yield (0, webdriverio_1.remote)(WDIO_PARAMS));
            updateWdioBrowser(driver);
        });
    });
    it('should find texts via new endpoint', function () {
        return __awaiter(this, void 0, void 0, function* () {
            yield driver.$('~Login Screen').waitForExist({ timeout: 5000 });
            const ocr = yield driver.getOcrText();
            const lines = ocr.lines.filter((l) => l.text.includes('Login Screen'));
            (0, earl_1.expect)(lines).toHaveLength(1);
            (0, earl_1.expect)(lines[0].confidence).toBeGreaterThan(0);
            (0, earl_1.expect)(lines[0].bbox.x0).toBeGreaterThan(0);
            (0, earl_1.expect)(lines[0].bbox.x1).toBeGreaterThan(0);
            (0, earl_1.expect)(lines[0].bbox.y0).toBeGreaterThan(0);
            (0, earl_1.expect)(lines[0].bbox.y1).toBeGreaterThan(0);
        });
    });
    it('should add a new context to the context list', function () {
        return __awaiter(this, void 0, void 0, function* () {
            const contexts = yield driver.getContexts();
            (0, earl_1.expect)(contexts).toInclude(__1.OCR_CONTEXT);
        });
    });
    describe('when in the ocr context', function () {
        before(function () {
            return __awaiter(this, void 0, void 0, function* () {
                yield driver.switchContext(__1.OCR_CONTEXT);
            });
        });
        after(function () {
            return __awaiter(this, void 0, void 0, function* () {
                yield driver.switchContext('NATIVE_APP');
            });
        });
        it('should get the ocr text in xml format', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const source = yield driver.getPageSource();
                (0, earl_1.expect)(source).toEqual(earl_1.expect.regex(/<OCR>/));
                (0, earl_1.expect)(source).toEqual(earl_1.expect.regex(/<words>/));
                (0, earl_1.expect)(source).toEqual(earl_1.expect.regex(/<lines>/));
            });
        });
        it('should find single ocr elements by xpath', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const el = yield driver.$('//lines/item[contains(text(), "Login Screen")]');
            });
        });
        it('should find multiple ocr elements by xpath', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const els = yield driver.$$('//lines/item[contains(text(), "Demo")]');
                (0, earl_1.expect)(els.length).toBeGreaterThan(4);
            });
        });
        it('should not find elements that do not exist', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const els = yield driver.$$('//item[contains(text(), "ZZZZ")]');
                (0, earl_1.expect)(els.length).toEqual(0);
            });
        });
        describe('with an element', function () {
            let el;
            const targetSize = {
                width: earl_1.expect.closeTo(95, 5),
                height: earl_1.expect.closeTo(16, 5)
            };
            const targetLoc = {
                x: earl_1.expect.closeTo(15, 5),
                y: earl_1.expect.closeTo(170, 5),
            };
            before(function () {
                return __awaiter(this, void 0, void 0, function* () {
                    el = yield driver.$('//lines/item[contains(text(), "Login Screen")]');
                });
            });
            it('should get displayed', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    (0, earl_1.expect)(yield el.isDisplayed()).toEqual(true);
                });
            });
            it('should get the size', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    (0, earl_1.expect)(yield el.getSize()).toEqual(targetSize);
                });
            });
            it('should get the location', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    (0, earl_1.expect)(yield el.getLocation()).toEqual(targetLoc);
                });
            });
            it('should get the rect', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    (0, earl_1.expect)(yield driver.getElementRect(el.elementId)).toEqual(Object.assign(Object.assign({}, targetSize), targetLoc));
                });
            });
            it('should get the text', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    (0, earl_1.expect)(yield el.getText()).toEqual(earl_1.expect.regex(/Login Screen/));
                });
            });
            it('should get the confidence attribute', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    const confidence = parseInt(yield el.getAttribute('confidence'), 10);
                    (0, earl_1.expect)(confidence).toBeGreaterThan(50);
                });
            });
            it('should click via w3c actions', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield el.click();
                    yield driver.switchContext('NATIVE_APP');
                    const username = yield driver.$('~username');
                    yield username.waitForExist({ timeout: 2000 });
                    yield driver.back();
                });
            });
        });
    });
    after(function () {
        return __awaiter(this, void 0, void 0, function* () {
            if (driver) {
                yield driver.deleteSession();
            }
        });
    });
});
