// appium-altunity-plugin/build/commands/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute_setComponentProperty = exports.execute_setTimeScale = exports.execute_getTimeScale = exports.execute = exports.EXECUTE_PREFIX = void 0;
const driver_1 = require("appium/driver");
exports.EXECUTE_PREFIX = 'unity:';
async function execute(next, _, script, args) {
    const cmdMap = {
        setTimeScale: this.execute_setTimeScale,
        getTimeScale: this.execute_getTimeScale,
        setProperty: this.execute_setComponentProperty,
    };
    return await this.unityContextGuard(next, async () => {
        const command = script.replace(new RegExp(`^${exports.EXECUTE_PREFIX} ?`), '');
        if (!Object.keys(cmdMap).includes(command)) {
            throw new driver_1.errors.InvalidArgumentError(`There is no '${exports.EXECUTE_PREFIX}${command}' command`);
        }
        return await cmdMap[command].bind(this)(...args);
    });
}
exports.execute = execute;
async function execute_getTimeScale() {
    return await this.client.getTimeScale();
}
exports.execute_getTimeScale = execute_getTimeScale;
async function execute_setTimeScale(timeScale) {
    await this.client.setTimeScale(timeScale);
}
exports.execute_setTimeScale = execute_setTimeScale;
async function execute_setComponentProperty(args) {
    const next = async () => {
        throw new Error(`Cannot call set component property without valid element id`);
    };
    return await this.unityElementGuard(next, args.elementId, async (el) => {
        return await el.element.setComponentProperty(args.component, args.property, '', args.value);
    });
}
exports.execute_setComponentProperty = execute_setComponentProperty;
