// appium-altunity-plugin/build/commands/source.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPageSource = void 0;
const xml_js_1 = __importDefault(require("xml-js"));
const element_1 = require("./element");
function hierarchyElementToXmlJs(element) {
    return {
        type: 'element',
        name: element.element.name.replace(/[^a-zA-Z0-9_\-\.]/g, '_'),
        attributes: { ...element.element.asUnityObject() },
        elements: element.children.map(hierarchyElementToXmlJs)
    };
}
function hierarchyToXmlJs(hierarchy) {
    return {
        elements: [{
                type: 'element',
                name: 'Unity',
                elements: hierarchy.map(hierarchyElementToXmlJs)
            }]
    };
}
async function getPageSource(next) {
    return await this.unityContextGuard(next, async () => {
        const allEls = await this.client.findAllObjects();
        // pre-emptively add all the elements we find in the cache
        for (const el of allEls) {
            const unityEl = new element_1.UnityElement(el);
            this.unityElements[unityEl.id] = unityEl;
        }
        const hierarchy = await this.client.getHierarchy(allEls);
        const hierarchyJs = hierarchyToXmlJs(hierarchy);
        return xml_js_1.default.js2xml(hierarchyJs, { spaces: 4 });
    });
}
exports.getPageSource = getPageSource;
