// appium-dashboard/lib/plugin/session-timeout-tracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionTimeoutTracker = void 0;
class SessionTimeoutTracker {
    constructor(options) {
        this.options = options;
    }
    start() {
        if (!this.timer) {
            this.lastCommandRecievedTime = new Date();
            this.timer = setInterval(() => {
                this.checkForSessionTimeout();
            }, this.options.pollingInterval);
        }
    }
    tick() {
        this.lastCommandRecievedTime = new Date();
    }
    stop() {
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = null;
        }
    }
    checkForSessionTimeout() {
        const timediffInSeconds = Math.round((new Date().getTime() - this.lastCommandRecievedTime.getTime()) / 1000);
        if (timediffInSeconds > this.options.timeout) {
            this.stop();
            this.options.timeoutCallback(this.options.timeout);
        }
    }
}
exports.SessionTimeoutTracker = SessionTimeoutTracker;
