// appium-altunity-plugin/build/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AltElement = exports.AltKeyCode = exports.AltBy = exports.AltUnityClient = exports.AltUnityPlugin = exports.UNITY_CONTEXT = void 0;
const plugin_1 = require("appium/plugin");
const client_1 = __importStar(require("./client"));
exports.AltUnityClient = client_1.default;
Object.defineProperty(exports, "AltBy", { enumerable: true, get: function () { return client_1.AltBy; } });
Object.defineProperty(exports, "AltElement", { enumerable: true, get: function () { return client_1.AltElement; } });
Object.defineProperty(exports, "AltKeyCode", { enumerable: true, get: function () { return client_1.AltKeyCode; } });
const commands_1 = require("./commands");
const EXTRA_CAPS = {
    altUnityHost: { presence: true },
    altUnityPort: { presence: true },
};
exports.UNITY_CONTEXT = 'UNITY';
class AltUnityPlugin extends plugin_1.BasePlugin {
    constructor(name) {
        super(name);
        this.unityElements = {};
        this.getPageSource = commands_1.getPageSource;
        this.getCurrentContext = commands_1.getCurrentContext;
        this.getContexts = commands_1.getContexts;
        this.setContext = commands_1.setContext;
        this.unityElementGuard = commands_1.unityElementGuard;
        this.click = commands_1.click;
        this.elementDisplayed = commands_1.elementDisplayed;
        this.getLocation = commands_1.getLocation;
        this.getElementRect = commands_1.getElementRect;
        this.getText = commands_1.getText;
        this.getAttribute = commands_1.getAttribute;
        this.getProperty = commands_1.getProperty;
        this.findElement = commands_1.findElement;
        this.findElements = commands_1.findElements;
        this._find = commands_1._find;
        this._findWithAltBy = commands_1._findWithAltBy;
        this.performActions = commands_1.performActions;
        this._performKeyActions = commands_1._performKeyActions;
        this._performPointerActions = commands_1._performPointerActions;
        this.getUrl = commands_1.getUrl;
        this.setUrl = commands_1.setUrl;
        this.execute = commands_1.execute;
        this.execute_getTimeScale = commands_1.execute_getTimeScale;
        this.execute_setTimeScale = commands_1.execute_setTimeScale;
        this.execute_setComponentProperty = commands_1.execute_setComponentProperty;
        this.client = new client_1.default({ log: this.logger });
        this.isInUnityContext = false;
    }
    shouldAvoidProxy(method, route) {
        if (this.isInUnityContext) {
            return true;
        }
        if (route.match(/\/contexts?$/)) {
            return true;
        }
        return false;
    }
    async createSession(next, driver) {
        if (driver.desiredCapConstraints) {
            driver.desiredCapConstraints = { ...driver.desiredCapConstraints, ...EXTRA_CAPS };
        }
        else {
            driver.desiredCapConstraints = EXTRA_CAPS;
        }
        // run the actual driver's create session now create session is weird and returns webdriver
        // protocol response, not a raw result, so we have to handle that
        const { value, error } = await next();
        if (error) {
            throw error;
        }
        const [id, caps] = value;
        this.platform = this.getPlatform(caps.platformName);
        // instantiate our unity client
        const { altUnityHost, altUnityPort } = caps;
        this.client = new client_1.default({ host: altUnityHost, port: altUnityPort, log: this.logger });
        await this.client.connect();
        return [id, caps];
    }
    getPlatform(platformName) {
        switch (platformName.toLowerCase()) {
            case 'mac': return "mac" /* Platform.Mac */;
            case 'windows': return "windows" /* Platform.Windows */;
            case 'linux': return "linux" /* Platform.Linux */;
            case 'ios': return "ios" /* Platform.iOS */;
            case 'android': return "android" /* Platform.Android */;
            default: throw new Error(`Can't find platform for '${platformName}'`);
        }
    }
    async unityContextGuard(next, fn) {
        if (!this.isInUnityContext) {
            return await next();
        }
        return await fn();
    }
    async getCachedScreenDims(driver) {
        if (!this.cachedScreenDims) {
            if (!driver.getWindowSize) {
                throw new Error(`Tried to get window size from driver but it doesn't support it`);
            }
            this.cachedScreenDims = await driver.getWindowSize();
        }
        return this.cachedScreenDims;
    }
    async deleteSession(next) {
        try {
            await this.client.disconnect();
        }
        finally {
            await next();
        }
    }
}
exports.AltUnityPlugin = AltUnityPlugin;
